### PR TITLE
SPARK-11326: Split networking in standalone mode

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

**Comment by jacek-lewandowski** [comment]
jenkins, test this please

**Comment by jacek-lewandowski** [comment]
@vanzin could you take a look?

**Comment by jacek-lewandowski** [comment]
jenkins, test this please

File: core/src/main/scala/org/apache/spark/Namespace.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+/** A simple trait that manages namespaces in Spark configuration properties and env variables */
+sealed private[spark] trait Namespace {
+  /** Namespace name for env variables */
+  def envNamespace: String
+
+  /** Namespace name for conf property */
+  def confNamespace: String
+
+  /** Adds this namespace to the given env variable name. Assuming the given `envKey` is
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/Namespace.scala]: this doesn't follow the multi-line scaladoc format used in Spark.

    /**
     * Text.
     */

Same in other places.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 2]
```diff
@@ -483,7 +482,7 @@ private[spark] object SecurityManager {
   val SPARK_AUTH_SECRET_CONF: String = "spark.authenticate.secret"
   // This is used to set auth secret to an executor's env variable. It should have the same
   // value as SPARK_AUTH_SECRET_CONF set in SparkConf
-  val ENV_AUTH_SECRET = "_SPARK_AUTH_SECRET"
+  val ENV_AUTH_SECRET = "SPARK_AUTH_SECRET"
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: Why this change? The `_` is normally used to indicate something that users should not set directly, and that's the case with this env variable.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 3]
```diff
@@ -364,6 +364,12 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
   /** Does the configuration contain a given parameter? */
   def contains(key: String): Boolean = settings.containsKey(key)
 
+  /** Returns a copy of this configuration with all the properties at `spark.srcNamespace.xxx`
+    * copied to `spark.xxx` */
+  def fromNamespace(namespace: Namespace): SparkConf = {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: `Namespace` is `private[spark]` but this method is public. Scala should have printed a warning about this. But that means this method should be `private[spark]`.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 4]
```diff
@@ -364,6 +364,12 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
   /** Does the configuration contain a given parameter? */
   def contains(key: String): Boolean = settings.containsKey(key)
 
+  /** Returns a copy of this configuration with all the properties at `spark.srcNamespace.xxx`
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: nit: param name is `namespace` not `srcNamespace`.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 5]
```diff
@@ -44,6 +44,7 @@ import org.apache.hadoop.mapreduce.{InputFormat => NewInputFormat, Job => NewHad
 import org.apache.hadoop.mapreduce.lib.input.{FileInputFormat => NewFileInputFormat}
 
 import org.apache.mesos.MesosNativeLibrary
+import org.apache.spark.SecurityManager._
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: nit: import order

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 6]
```diff
@@ -379,6 +380,16 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   try {
     _conf = config.clone()
+
+    // if a secret key has been set in an env variable, use it
+    for (ns <- Seq(Namespace.Blank, Namespace.Application, Namespace.Submission)) {
+      sys.env.get(ns.inEnv(ENV_AUTH_SECRET))
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Now that I see the methods being used, the names sound kinda weird. I'd prefer `ns.envName` and `ns.confName`.

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 7]
```diff
@@ -23,9 +23,10 @@ import scala.reflect.ClassTag
 import scala.util.{Failure, Success}
 
 import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SecurityManager.{SPARK_AUTH_SECRET_CONF, ENV_AUTH_SECRET}
 
 import org.apache.spark.rpc.{RpcEndpointRef, RpcAddress, RpcEnv, ThreadSafeRpcEndpoint}
-import org.apache.spark.{Logging, SecurityManager, SparkConf}
+import org.apache.spark.{Namespace, Logging, SecurityManager, SparkConf}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: nit: import order

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 8]
```diff
@@ -23,9 +23,10 @@ import scala.reflect.ClassTag
 import scala.util.{Failure, Success}
 
 import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SecurityManager.{SPARK_AUTH_SECRET_CONF, ENV_AUTH_SECRET}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: nit: import order

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 9]
```diff
@@ -80,11 +81,30 @@ private class ClientEndpoint(
...
+          .map(_.inConf(SPARK_AUTH_SECRET_CONF))
+        val sparkJavaOpts = Utils.sparkJavaOpts(conf, key => !propsToExclude.contains(key))
         val javaOpts = sparkJavaOpts ++ extraJavaOpts
-        val command = new Command(mainClass,
-          Seq("{{WORKER_URL}}", "{{USER_JAR}}", driverArgs.mainClass) ++ driverArgs.driverOptions,
-          sys.env, classPathEntries, libraryPathEntries, javaOpts)
+
+        // we want to transport secrets only from submission and app namespaces; they will be
+        // transported as environment variables though
+        val envSecrets = Set(Namespace.Submission, Namespace.Application).flatMap { ns =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: I think used `Seq` in these places would be better, since it's faster and you don't really need the functionality of a `Set` here.

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 10]
```diff
@@ -80,11 +81,30 @@ private class ClientEndpoint(
...
+          sys.env.get(ns.inEnv(ENV_AUTH_SECRET))
+            .orElse(conf.getOption(ns.inConf(SPARK_AUTH_SECRET_CONF)))
+            .map(v => ns.inEnv(ENV_AUTH_SECRET) -> v)
+        }
+
+        val command = new Command(
+          mainClass = mainClass,
+          arguments = Seq("{{WORKER_URL}}", "{{USER_JAR}}", driverArgs.mainClass)
+            ++ driverArgs.driverOptions,
+          environment = sys.env ++ envSecrets - ENV_AUTH_SECRET,
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: what's the precedence order here? can you add parentheses to make that explicit (and make sure the code does what's intended)?

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 11]
```diff
@@ -208,6 +228,9 @@ private class ClientEndpoint(
  * Executable utility for starting and terminating drivers inside of a standalone cluster.
  */
 object Client {
+  val SYSTEM_NAME = "driverClient"
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: these should not be public (since `Client` is a public class).

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 12]
```diff
@@ -226,12 +249,29 @@ object Client {
     conf.set("akka.loglevel", driverArgs.logLevel.toString.replace("WARN", "WARNING"))
     Logger.getRootLogger.setLevel(driverArgs.logLevel)
 
-    val rpcEnv =
-      RpcEnv.create("driverClient", Utils.localHostName(), 0, conf, new SecurityManager(conf))
+    // if a secret key has been set in an env variable, use it
+    sys.env.get(ENV_AUTH_SECRET).foreach(conf.set(SPARK_AUTH_SECRET_CONF, _))
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: This should never happen. The user is not supposed to set that env variable.

File: core/src/main/scala/org/apache/spark/deploy/Client.scala [hunk 13]
```diff
@@ -226,12 +249,29 @@ object Client {
...
+
+    // if a secret key for submission has been set in an env variable, use it
+    sys.env.get(Namespace.Submission.inEnv(ENV_AUTH_SECRET))
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: As above, this code should not be added.

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 14]
```diff
@@ -17,17 +17,16 @@
...
 import scala.util.control.NonFatal
 
-import org.apache.spark.{Logging, SparkConf}
-import org.apache.spark.deploy.{ApplicationDescription, ExecutorState}
 import org.apache.spark.deploy.DeployMessages._
 import org.apache.spark.deploy.master.Master
+import org.apache.spark.deploy.{ApplicationDescription, ExecutorState}
 import org.apache.spark.rpc._
 import org.apache.spark.util.{RpcUtils, ThreadUtils, Utils}
+import org.apache.spark.{Logging, SecurityManager, SparkConf, SparkContext, Namespace}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: nit: import order.

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 15]
```diff
@@ -99,8 +98,10 @@ private[spark] class AppClient(
               return
             }
             logInfo("Connecting to master " + masterAddress.toSparkURL + "...")
-            val masterRef =
-              rpcEnv.setupEndpointRef(Master.SYSTEM_NAME, masterAddress, Master.ENDPOINT_NAME)
+            val masterRef = rpcEnv.setupEndpointRef(
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: I see changes like this in lots of places. These are not arguments with default values, so calling them like this is completely unnecessary and just adds noise. Can you please revert changes like this from your patch?

**Comment by jacek-lewandowski** [comment]
@vanzin thanks a lot for your review. Of course, I'll apply all those rename/import/style changes. But the thing about env variable needs explanation - actually the user is supposed to provide a secret key in env variable. Especially, it is much more secure than providing it in a command line of spark-submit (as you know). Now that the network communication layers are separated, the user would have to provide "credentials" in order to submit the application.

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 16]
```diff
@@ -303,3 +305,25 @@ private[spark] class AppClient(
   }
 
 }
+
+object AppClient {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: This should not be public.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 17]
```diff
@@ -30,7 +30,7 @@ import scala.util.Random
 import org.apache.hadoop.fs.Path
 
 import org.apache.spark.rpc._
-import org.apache.spark.{Logging, SecurityManager, SparkConf, SparkException}
+import org.apache.spark.{Logging, SecurityManager, SparkConf, SparkException, Namespace}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: nit: import order

**Comment by vanzin** [comment]
> But the thing about env variable needs explanation - actually the user is supposed to provide a secret key in env variable.

I haven't finished my review yet, but I was going to get to that. That change is really not necessary.

For the secret needed to connect to the Spark Master, that can be in the configuration file. Everybody needs to share the same secret, so there's no problem in it being there. Until the Master can do proper authentication, we'll have to live with that fact.

For the per-app secret, it should be auto-generated, like in YARN. The user should not need to provide their own secret every time they launch their app.

So you don't really have to set any env variables. The env variable is just a way to get the secret from the Worker to the executor without using the command line.

BTW, I think this is a step in the right direction, but because of the above, this is not really adding any extra security. Because all Spark-launched processes will run as the same user, it's possible for any malicious executor to just read other app's secrets from the Worker's memory (or directly from other running executors' environment in /proc).

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 18]
```diff
@@ -47,11 +47,15 @@ import org.apache.spark.util.{ThreadUtils, SignalLogger, Utils}
...
 private[deploy] class Master(
     override val rpcEnv: RpcEnv,
-    address: RpcAddress,
+    val rpcSubmissionEnv: RpcEnv,
     webUiPort: Int,
+    restSubmissionPort: Int,
     val securityMgr: SecurityManager,
     val conf: SparkConf)
-  extends ThreadSafeRpcEndpoint with Logging with LeaderElectable {
+  extends ThreadSafeRpcEndpoint with Logging with LeaderElectable { master =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: nit: in other Spark code, `master =>` is on its own line.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 19]
```diff
@@ -126,27 +131,92 @@ private[deploy] class Master(
...
+
+      case msg @ MasterChangeAcknowledged(appId) =>
+        master.self.send(msg)
+
+      case msg @ UnregisterApplication(applicationId) =>
+        master.self.send(msg)
+    }
+
+    override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+      case msg @ RequestSubmitDriver(description) =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: since you're not really using any of the arguments, can you just say `case msg: RequestSubmitDriver =>` here? I'm not sure Scala allows that for case classes but that would be much cleaner.

If you can't, then you can just use placeholders (`_`) for the arguments.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 20]
```diff
@@ -126,27 +131,92 @@ private[deploy] class Master(
...
-  private val restServerEnabled = conf.getBoolean("spark.master.rest.enabled", true)
-  private var restServer: Option[StandaloneRestServer] = None
-  private var restServerBoundPort: Option[Int] = None
+  private val restSubmissionServerEnabled = conf.getBoolean("spark.master.rest.enabled", true)
+  private var restSubmissionServer: Option[StandaloneRestServer] = None
+  private var restSubmissionServerBoundPort: Option[Int] = None
+
+  private var rpcSubmissionEndpointRef: Option[RpcEndpointRef] = None
+
+  object RPCSubmissionEndpoint extends ThreadSafeRpcEndpoint {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: it's not really clear: why do you need this? why can't you just use the Master itself, since it already handles all these messages?

At the very least, this needs a comment (in the code) explaining that. Also, nested classes are generally declared at the bottom of the enclosing class.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 21]
```diff
@@ -1051,32 +1133,61 @@ private[deploy] class Master(
 
 private[deploy] object Master extends Logging {
   val SYSTEM_NAME = "sparkMaster"
+  val SUBMISSION_SYSTEM_NAME = "sparkSubmission"
   val ENDPOINT_NAME = "Master"
+  val SUBMISSION_ENDPOINT_NAME = "Master"
+
+  val DefaultRpcPort = 7077
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: nit: constants are DEFINED_IN_CAPS. Just look a few lines above.

File: core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala [hunk 22]
```diff
@@ -24,30 +24,35 @@ import org.apache.spark.util.{IntParam, Utils}
...
-  if (System.getenv("SPARK_MASTER_HOST") != null) {
-    host = System.getenv("SPARK_MASTER_HOST")
-  }
-  if (System.getenv("SPARK_MASTER_PORT") != null) {
-    port = System.getenv("SPARK_MASTER_PORT").toInt
-  }
-  if (System.getenv("SPARK_MASTER_WEBUI_PORT") != null) {
-    webUiPort = System.getenv("SPARK_MASTER_WEBUI_PORT").toInt
-  }
+  sys.env.get("SPARK_MASTER_HOST").foreach(v => host = v)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala]: nit: `foreach { v => ... }` (we use `()` only when using `_`)

File: core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala [hunk 23]
```diff
@@ -24,30 +24,35 @@ import org.apache.spark.util.{IntParam, Utils}
...
+  sys.env.get("SPARK_MASTER_PORT").foreach(v => port = v.toInt)
+  sys.env.get("SPARK_MASTER_WEBUI_PORT").foreach(v => webUiPort = v.toInt)
+
+  sys.env.get("SPARK_MASTER_SUBMISSION_RPC_PORT").orElse(sys.env.get("SPARK_MASTER_PORT"))
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala]: Why add a new name for the env variable? Is `SPARK_MASTER_PORT` gonna change meaning because of this change?

Changing configuration is kinda bad because you still need to keep the old code, and need to educate users on the new values. So unless there's a very good reason for the change, I don't see the point.

File: core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala [hunk 24]
```diff
@@ -24,30 +24,35 @@ import org.apache.spark.util.{IntParam, Utils}
...
   propertiesFile = Utils.loadDefaultSparkProperties(conf, propertiesFile)
 
-  if (conf.contains("spark.master.ui.port")) {
-    webUiPort = conf.get("spark.master.ui.port").toInt
-  }
+  conf.getOption("spark.master.host").foreach(v => host = v)
+  conf.getOption("spark.master.ui.port").foreach(v => webUiPort = v.toInt)
+  conf.getOption("spark.master.rpc.port").foreach(v => port = v.toInt)
+  conf.getOption("spark.master.submission.rest.port").foreach(v => restSubmissionPort = v.toInt)
+  conf.getOption("spark.master.submission.rpc.port").foreach(v => rpcSubmissionPort = v.toInt)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala]: It's really confusing to have conf split like this. You should merge the two pieces of code setting `rpcSubmissionPort` into one statement.

I'd also recommend declaring `vals` and using `getOrElse` instead of `foreach`.

File: core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala [hunk 25]
```diff
@@ -68,6 +73,14 @@ private[master] class MasterArguments(args: Array[String], conf: SparkConf) {
       webUiPort = value
       parse(tail)
 
+    case ("--rpc-submission-port") :: IntParam(value) :: tail =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala]: I'd rather not add command line parameters. Just keep everything in the conf file.

File: core/src/main/scala/org/apache/spark/deploy/worker/CommandUtils.scala [hunk 26]
```diff
@@ -22,8 +22,8 @@ import java.io.{File, FileOutputStream, InputStream, IOException}
 import scala.collection.JavaConverters._
 import scala.collection.Map
 
-import org.apache.spark.Logging
-import org.apache.spark.SecurityManager
+import org.apache.spark.{Namespace, Logging, SecurityManager}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/worker/CommandUtils.scala]: nit: import order

File: core/src/main/scala/org/apache/spark/deploy/worker/CommandUtils.scala [hunk 27]
```diff
@@ -117,4 +114,15 @@ object CommandUtils extends Logging {
       }
     }.start()
   }
+
+  val SystemPropertyPattern = """^-D(.+?)=(.*)$""".r
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/worker/CommandUtils.scala]: nit: either ALL_CAPS or camelCase. We don't declare variables with names like that in Spark.

**Comment by markhamstra** [comment]
Please see my JIRA comment.

**Comment by vanzin** [comment]
I stopped when I got to the tests; overall, I share the same concerns Mark mentioned on the bug. I don't really know what the goals of standalone mode are in the long run, but if your concern is security, there's a lot of much harder work needed to get that working.

File: core/src/main/scala/org/apache/spark/deploy/Client.scala (hunk 8 above)
Comment (jacek-lewandowski) [inline core/src/main/scala/org/apache/spark/deploy/Client.scala]: Weird that IDE didn't take care of that :(

**Comment by vanzin** [comment]
BTW you probably want @andrewor14 to take a look at this stuff.

**Comment by jacek-lewandowski** [comment]
@vanzin I've changed it a little bit. The master doesn't create two endpoints now but different users can be used with a single endpoint. So basically the user can plug his own password authenticator if he wants to which will be used to authenticate Spark users with the master. 

To authenticate workers with the master, and clients when no special password authenticator is plugged in, the default SASL user is used (as it was used so far). When the user wants to enable more sophisticated authentication, he simply needs to provide some authenticator or just set different secrets for the default user {{sparkSaslUser}} (which is used by the workers) and other users (the default authenticator allows to set multiple secrets like {{spark.authenticate.secrets.<username>=password}} while the default password is set as usual with {{spark.authenticate.secret}}). In this case, the users should just do not need to know the default user password. Obviously, this only works for Netty RPC. The whole original behaviour when no-secret or a single shared secret is provided remains unchanged.

The last PR is not polished, but it removes most of the changes from the master. The remaining thing is authorisation - determining who is communicating with the master and authorising the requests (for example, only a default user can exchange worker messages, or only the application owner can remove his application).

**Comment by jacek-lewandowski** [comment]
jenkins, test this please

**Comment by jacek-lewandowski** [comment]
Rebased and more reasonably split into commits.

**Comment by jacek-lewandowski** [comment]
jenkins, test this please

**Comment by HyukjinKwon** [comment]
Maybe we might have to correct the title just like the others, `[SPARK-XXXX][SQL]` (this is described in https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark).
Also, it would be better if the description of this PR is updated, It is empty.

**Comment by tnachen** [comment]
This has been stale for a while, we should close this if there is no update here.

**Comment by HyukjinKwon** [comment]
+1 for closing.

**Comment by jacek-lewandowski** [comment]
Closing, although I don't really understand the reason why it wasn't accepted.

**Comment by HyukjinKwon** [comment]
I ~guess~ the reason it wasn't accepted and reviewed due to the lack of man power of committers or the lack of appealing or because it simply just got stale. At least, I said +1 for this reason.

I ~think~ it is okay to reopen, fix the codes to pass the tests, then keep pining reviewers after filling the PR description with a proper title according to the glide lines and an email to dev mailing list, yelling to the public to explain why this fix is right and nedded. Usually, it is really hard for a lot of changes to be merged quickly because it is hard to review. For this reason, I also have several old PRs but I am trying to explain and keep pinging with rebasing everytime they get conflicts

This one is not really in my area but if you strongly think this is the right fix and the important change,  I'd keep open this and then try to persuade reviwers with keeping pininging.
